packages/frontend-defaults/src/resolution.ts

```diff
@@ -16,30 +16,98 @@
 
 import { Config } from '@backstage/config';
 import { stringifyError } from '@backstage/errors';
-import { FrontendFeature } from '@backstage/frontend-app-api';
+import {
+  FrontendFeature,
+  FrontendFeatureLoader,
+} from '@backstage/frontend-plugin-api';
 import { CreateAppFeatureLoader } from './createApp';
+// eslint-disable-next-line @backstage/no-relative-monorepo-imports
+import { isInternalFrontendFeatureLoader } from '../../frontend-plugin-api/src/wiring/createFrontendFeatureLoader';
 
 /** @public */
 export async function resolveAsyncFeatures(options: {
   config: Config;
-  features?: (FrontendFeature | CreateAppFeatureLoader)[];
+  features?: (
+    | FrontendFeature
+    | FrontendFeatureLoader
+    | CreateAppFeatureLoader
+  )[];
 }): Promise<{ features: FrontendFeature[] }> {
-  const features = [];
-  for (const entry of options.features ?? []) {
-    if ('load' in entry) {
+  const features: (FrontendFeature | FrontendFeatureLoader)[] = [];
+
+  // Separate deprecated CreateAppFeatureLoader elements from the frontend features,
+  // and manage the deprecated elements first.
+  for (const item of options?.features ?? []) {
+    if ('load' in item) {
       try {
-        const result = await entry.load({ config: options.config });
+        const result = await item.load({ config: options.config });
         features.push(...result.features);
       } catch (e) {
         throw new Error(
-          `Failed to read frontend features from loader '${entry.getLoaderName()}', ${stringifyError(
+          `Failed to read frontend features from loader '${item.getLoaderName()}', ${stringifyError(
             e,
           )}`,
         );
       }
     } else {
-      features.push(entry);
+      features.push(item);
+    }
+  }
+
+  const loadedFeatures: FrontendFeature[] = [];
+  const alreadyMetFeatureLoaders: FrontendFeatureLoader[] = [];
+  const maxRecursionDepth = 5;
+
+  async function applyFeatureLoaders(
+    featuresOrLoaders: (FrontendFeature | FrontendFeatureLoader)[],
+    recursionDepth: number,
+  ) {
+    if (featuresOrLoaders.length === 0) {
+      return;
+    }
+
+    for (const featureOrLoader of featuresOrLoaders) {
+      if (isBackstageFeatureLoader(featureOrLoader)) {
+        if (alreadyMetFeatureLoaders.some(l => l === featureOrLoader)) {
```

Comment:
`.includes` would work here instead, or even better make it a `Set` and use `.has`. Merging as is though as this is minor